pkg/metrics/policystatusmetrics/policystatusmetrics.go

```diff
@@ -0,0 +1,91 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Tetragon
+
+package policystatusmetrics
+
+import (
+	"context"
+
+	"github.com/cilium/tetragon/api/v1/tetragon"
+	"github.com/cilium/tetragon/pkg/logger"
+	"github.com/cilium/tetragon/pkg/metrics/consts"
+	"github.com/cilium/tetragon/pkg/observer"
+	"github.com/cilium/tetragon/pkg/sensors"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+type policyStatusCollector struct {
+	descriptor    *prometheus.Desc
+	sensorManager *sensors.Manager
+}
+
+func InitMetrics(registry *prometheus.Registry) {
+	registry.MustRegister(newPolicyStatusCollector(observer.GetSensorManager()))
+}
+
+// This metric collector converts the output of ListTracingPolicies into a few
+// gauges metrics on collection. Thus, it needs a sensor manager to query.
+func newPolicyStatusCollector(sensorManager *sensors.Manager) *policyStatusCollector {
+	return &policyStatusCollector{
+		descriptor: prometheus.NewDesc(
+			prometheus.BuildFQName(consts.MetricsNamespace, "", "tracingpolicy_loaded"),
+			"The number of loaded tracing policy by status.",
+			[]string{"status"}, nil,
+		),
+		sensorManager: sensorManager,
+	}
+}
+
+func (c *policyStatusCollector) Describe(ch chan<- *prometheus.Desc) {
+	ch <- c.descriptor
+}
+
+func (c *policyStatusCollector) Collect(ch chan<- prometheus.Metric) {
+	if c.sensorManager == nil {
+		logger.GetLogger().Debug("failed retrieving the sensor manager: manager is nil")
+		return
+	}
+	list, err := c.sensorManager.ListTracingPolicies(context.Background())
+	if err != nil {
+		logger.GetLogger().WithError(err).Warn("error listing tracing policies to collect policies status")
+		return
+	}
+
+	metrics := c.convertStatusToMetrics(list.Policies)
+	for _, metric := range metrics {
+		ch <- metric
+	}
+}
+
+func (c *policyStatusCollector) convertStatusToMetrics(policiesStatus []*tetragon.TracingPolicyStatus) []prometheus.Metric {
+	loadErrorCount, errorCount, disabledCount, enabledCount := 0, 0, 0, 0
+
+	// This for loop assumes that a tracing policy can only be in one of those
+	// states exclusively: load error, error, disabled or enabled.
+	//
+	// Meaning that a tracing policy having a load error cannot also have an
+	// error, and that anything with a load error or just an error cannot be
+	// enabled or disabled. See unit test.
```

Comment:
Is there any sort of state machine for tracing policies? This comment suggests that there is an implicit but not an explicit one, but I'm not familiar with the policy manager code.

It took me a moment to understand what this comment means: on the struct level a policy is always either enabled or disabled, and can have both errors. But here we count each policy in exactly one metric - meaning, the order of the `if`s matters a lot, i.e. determines the state of the policy. Does it sound right? Are there any combinations of these fields that would mean a bug in the policy manager?

Reply:
I would say no, there is no state machine. This is expressed in the unit test, I'm not sure there is something to prevent that a policy is enabled and has an error had the same time, I didn't check. For load_error I know that this is an exclusive state, and enabled/disabled is by nature exclusive.

So I would say most of this should be correct, and I think it's better for the user if we assume a policy is in only one state at a time and give the error state over instead of a confusing "enabled/error" if that's even possible.

Reply:
Ok, thanks. Could you mention in the comment that the state is determined here by the order of `if`s? I.e. on the struct level a policy has between 1 and 3 "states" (it's always enabled or disabled and you can have 0-2 errors), but here it gets mapped into 1 state.

Reply:
> > Is there any sort of state machine for tracing policies? This comment suggests that there is an implicit but not an explicit one, but I'm not familiar with the policy manager code.

> I would say no, there is no state machine.

I'd say there is an implicit state machine for policies. It is embedded in the "collection" structure. For example, if a collection's state is LOADED_ERROR it means that it holds no resources so we can just replace it in the collections map.

Reply:
Regarding this comment and Djalal output on `tetra` output, I put that state part in a separate function (that we could move somewhere else) that I can reuse (in tetra), and changed everything from `status` to `state` to avoid confusion (and I think this discussion highlighted that state makes more sense).